### PR TITLE
Refactor the code dealing with database into `BackendManager`

### LINKED ISSUE #3530: Refactor and abstract a database backend manager class
This class should be responsible for interacting directly with the database without going through the ORM. This is necessary for example to check database schema versions and  perform database migrations.

### LINKED ISSUE #3584: Unable to migrate database
We are trying to migrate a database from aiida 0.12.2 to aiida 1.0.
`verdi status` runs into an exception connecting to the postgres DB (just shown as an "x" for the postgres service).

We get the same exception when trying to migrate:
```
$ verdi database migrate
Warning: Migrating your database might take a while and is not reversible.
Warning: Before continuing, make sure you have completed the following steps:
Warning:
Warning:  1. Make sure you have no active calculations and workflows.
Warning:  2. If you do, revert the code to the previous version and finish running them first.
Warning:  3. Stop the daemon using `verdi daemon stop`
Warning:  4. Make a backup of your database and repository
Warning:
Warning: If you have completed the steps above and want to migrate profile "COFs", type MIGRATE NOW: MIGRATE NOW
Traceback (most recent call last):
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedColumn: column db_dbsetting.val does not exist
LINE 1: SELECT "db_dbsetting"."id", "db_dbsetting"."key", "db_dbsett...
                                                          ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/kdeeg/miniconda3/envs/aiida1/bin/verdi", line 11, in <module>
    sys.exit(verdi())
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/aiida/cmdline/commands/cmd_database.py", line 70, in database_migrate
    backend.migrate()
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/aiida/orm/implementation/django/backend.py", line 53, in migrate
    migrate_database()
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/aiida/backends/djsite/utils.py", line 124, in migrate_database
    validate_schema_generation()
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/aiida/backends/utils.py", line 78, in validate_schema_generation
    schema_generation_database = get_db_schema_generation().value
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/aiida/backends/utils.py", line 97, in get_db_schema_generation
    return manager.get(SCHEMA_GENERATION_KEY)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/aiida/backends/djsite/utils.py", line 50, in get
    setting = DbSetting.objects.filter(key=key).first()
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/models/query.py", line 567, in first
    objects = list((self if self.ordered else self.order_by('pk'))[:1])
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/models/query.py", line 250, in __iter__
    self._fetch_all()
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 899, in execute_sql
    raise original_exception
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 889, in execute_sql
    cursor.execute(sql, params)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/kdeeg/miniconda3/envs/aiida1/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column db_dbsetting.val does not exist
LINE 1: SELECT "db_dbsetting"."id", "db_dbsetting"."key", "db_dbsett...
```

I.e. the migration code assumes that a column `db_dbsetting.val` exists, which does not exist in our case.

@sphuber Is this a bug in the migrations?

Besides fixing the bug, I also don't know how to proceed here to complete the migration...

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #3530 and fixes #3584 

The code that acts directly with the database, outside of the ORM, for
example to check the database schema generation and version, or to get
data from the `DbSettings` table has been refactored. The ORM requires
the database environment to be loaded, but to do that the database
schema needs to be verified which also requires accessing the database.
This functionality was implemented mostly in free-standing functions and
separately for both backends. This is reorganized into a `BackendManager`
abstract class and implemented for Django and SqlAlchemy. This
centralizes the functionality required to verify database schema and
perform the migrations thereof.

### CODE REVIEW

**Comment by sphuber** [comment]
This is the refactoring work that I originally implemented in combination with the reset of the database schema generation. However, having been convinced to postpone this, I still think the refactoring of the backend utilities should go in now. All the groundwork for the schema reset for `v2.0.0` will then already be laid and all 1.0.* versions will have the correct messaging.

**Comment by ltalirz** [comment]
I'll leave this review to @giovannipizzi if that's ok

File: aiida/backends/sqlalchemy/utils.py [hunk 1]
```diff
@@ -153,24 +70,25 @@ def install_tc(session):
...
+    )
 
 
-def get_pg_tc(links_table_name,
-              links_table_input_field,
-              links_table_output_field,
-              closure_table_name,
-              closure_table_parent_field,
-              closure_table_child_field):
+def get_pg_tc(
```
Comment (giovannipizzi) [inline aiida/backends/sqlalchemy/utils.py]: Why do we still have transitive-closure code in our codebase?

Comment (sphuber) [reply aiida/backends/sqlalchemy/utils.py]: Because this is part of the original database schema and so is installed in the initial migration.

**Pull request merged**